File: rnamake/motif_state_search.py
import motif
import motif_state_tree
import base
import option
import priority_queue
import motif_state
import motif_state_selector
import motif_state_search_scorer

from collections import defaultdict

class MotifStateSearch(base.Base):
    __slots__ = [
        '_rm',
        '_queue',
        '_selector',
        '_scorer',
        '_solutions',
        '_test_node',
        '_no_more_solutions',
        'options'
    ]

    def __init__(self, rm):
        self._rm = rm
        self._queue = priority_queue.PriorityQueue()
        self._selector = motif_state_selector.default_selector(self._rm)
        self._scorer = motif_state_search_scorer.MTSS_Astar()
        self._solutions = []

        self._no_more_solutions = 0
        self.setup_options_and_constraints()
        #self.lookup = None
        #self.test_node = None

    def setup_options_and_constraints(self):
        options =     { 'sterics'        :  1,
                        'verbose'        :  0,
                        'frequency'      :  10,
                        'max_node_level'  : 20,
                        'max_steps'       : 100000000,
                        'max_solutions'   : 1,
                        'max_size'        : 100000000,
                        'min_size'        : 0,
                        'accept_score'    : 10.0}

        self.options = option.Options(options)

    def _get_local_variables(self):
        return self.option('accept_score'), self.option('max_node_level'), \
               self.option('sterics'), self.option('max_size'), self.option('min_size')

    def setup(self, start, start_ei, end, end_ei):
        start_n = MotifStateSearchNode(start, None, -1, -1, 10000000)
        self._test_node = MotifStateSearchNode(start, None, -1, -1, 1000000)
        self._scorer.set_target(end.get_end(end_ei))
        self._first_round(start_n, start_ei)

    def finished(self):
        if len(self._solutions) >= self.option("max_solutions") or \
            self.no_more_solutions == 1:
            return 1
        else:
            return 0

    def next(self):
        sol = self._search()
        if sol == None:
            self.no_more_solutions = 1
            return None
        self._solutions.append(sol)
        return sol

    def all(self):
        while not self._queue.empty():
            sol = self._search()
            self._solutions.append(sol)
            if len(self._solutions) >= self.option('max_solutions'):
                break
        return self._solutions

    def _first_round(self, start_n, start_ei):
        mlib_indexes = self._selector.get_connected_libs(-1)
        for ntype in mlib_indexes:
            motif_states = self._selector.get_motif_states(ntype)
            for ms in motif_states:
                new_ms = motif_state.Motif.copy(ms)
                motif_state.align_motif_state(start_n.state.get_end(start_ei), new_ms)
                self._test_node.update_ms(new_ms)
                score = self._scorer.score(self._test_node)
                new_node = MotifStateSearchNode(new_ms, start_n, start_ei, ntype, score)
                self._queue.put(new_node, score)

    def _search(self):
        accept_score, max_node_level, sterics, max_size, min_size = self._get_local_variables()
        best = 1000000000
        count = 0
        while not self._queue.empty():
            current = self._queue.get()
            score = self._scorer.accept_score(current)
            count += 1
            if score < best:
                best = score
                print best

            if count > 1000:
                break

            if score < accept_score:
                if current.size < min_size:
                    continue
                return MotifStateSearchSolution(current, score)

            if current.level+1 > max_node_level:
                continue

            mlib_indexes = self._selector.get_connected_libs(current.ntype)
            self._test_node.update_parent(current)
            for i, end in enumerate(current.state.iter_ends()):
                if i == 0:
                    continue
                for ntype in mlib_indexes:
                    motif_states = self._selector.get_motif_states(ntype)
                    for ms in motif_states:
                        motif_state.align_motif_state(end, ms)
                        self._test_node.update_ms(ms)
                        score = self._scorer.score(self._test_node)
                        #print score, current.score
                        if score > current.score:
                            continue
                        new_ms = motif_state.Motif.copy(ms)
                        new_node = MotifStateSearchNode(new_ms, current, i, ntype, score)
                        self._queue.put(new_node, score)

                        #print current.state.name, ms.name, score
        return None


class MotifStateSearchNode(object):
    def __init__(self, state, parent, parent_end_index, ntype, score):
        self.parent = parent
        self.parent_end_index = parent_end_index
        self.ss_score = state.score
        self.size =   state.num_res()
        self.ntype = ntype
        self.state = state
        if parent is None:
            self.level = 1
        else:
            self.level = self.parent.level + 1
            self.ss_score += self.parent.ss_score
            self.size += self.parent.size
        self.score = score

    def update_parent(self, parent):
        self.parent = parent
        self.ss_score = parent.ss_score
        self.size = parent.size

    def update_ms(self, ms):
        self.state = ms
        if self.parent is None:
            return
        self.level = self.parent.level + 1
        self.ss_score = self.state.score + self.parent.ss_score
        self.size = self.state.num_res() + self.parent.size


class MotifStateSearchSolution(object):
    def __init__(self, node, score):
        self.path = self._get_path(node)
        self.score = score

    def _get_path(self, node):
        path = []
        while node != None:
            path.append(node)
            node = node.parent
        return path[::-1][1:]

    def to_mst(self, rm):
        mst = motif_state_tree.MotifStateTree(rm)
        mst.set_sterics(0)
        for i, n in enumerate(self.path):
            new_ms = motif_state.Motif.copy(n.state, new_uuid=1)
            if i == 0:
                mst.add_state(new_ms)
            else:
                j = mst.add_state(new_ms, -1, n.parent_end_index)
                if j == -1:
                    raise ValueError("something went horribly wrong, cannot build solution")
        return mst

    def to_motif_tree(self, rm):
        return self.to_mst(rm).to_motif_tree()

File: rnamake/lib/RNAMake/src/motif_state_search/motif_state_search.cpp
//
//  motif_state_search.cpp
//  RNAMake
//
//  Created by Joseph Yesselman on 9/14/15.
//  Copyright (c) 2015 Joseph Yesselman. All rights reserved.
//

#include <climits>

#include "motif/motif_state.h"
#include "motif_state_search/motif_state_search.h"

void
MotifStateSearch::setup_options() {
    options_.add_option("sterics", true, OptionType::BOOL);
    options_.add_option("max_node_level", 100, OptionType::INT);
    options_.add_option("min_node_level", 0, OptionType::INT);
    options_.add_option("min_size", 0, OptionType::INT);
    options_.add_option("max_size", 1000000, OptionType::INT);
    options_.add_option("max_solutions", 1, OptionType::INT);
    options_.add_option("accept_score", 10, OptionType::FLOAT);
    options_.add_option("min_ss_score", 10000, OptionType::FLOAT);
    options_.add_option("max_steps", 1000000000, OptionType::FLOAT);
    options_.add_option("verbose", true, OptionType::BOOL);
    options_.add_option("return_best", false, OptionType::BOOL);
    options_.add_option("helix_end", false, OptionType::BOOL);
    
    //for making a movie
    options_.add_option("save_midpoints", false, OptionType::BOOL);
    options_.add_option("save_midpoints_file", "midpoints.dat", OptionType::STRING);
    options_.add_option("save_midpoints_freq", 100, OptionType::INT);
    
    options_.lock_option_adding();
    
    update_var_options();
}

void
MotifStateSearch::update_var_options() {
    sterics_        = options_.get_bool("sterics");
    min_size_       = options_.get_int("min_size");
    max_size_       = options_.get_int("max_size");
    max_solutions_  = options_.get_int("max_solutions");
    max_node_level_ = options_.get_int("max_node_level");
    min_node_level_ = options_.get_int("min_node_level");
    accept_score_   = options_.get_float("accept_score");
    min_ss_score_   = options_.get_float("min_ss_score");
    max_steps_      = options_.get_float("max_steps");
    verbose_        = options_.get_bool("verbose");
    helix_end_      = options_.get_bool("helix_end");
    
}

void
MotifStateSearch::setup(
    BasepairStateOP const & start,
    BasepairStateOP const & end) {
    
    auto start_n = _start_node(start);
    start_n->score(1000000000);
    queue_ =  MotifStateSearchNodeQueue();
    test_node_ = std::make_shared<MotifStateSearchNode>(*start_n);
    queue_.push(start_n);
    scorer_->set_target(end);
    no_more_solutions_ = 0;
    sol_count_ = 0;
}

void
MotifStateSearch::reset() {
    no_more_solutions_ = 0;
    sol_count_ = 0;
    
}

MotifStateSearchSolutionOP
MotifStateSearch::next() {

    auto sol = _search();
    if(sol == nullptr) {
        no_more_solutions_ = 1;
        return nullptr;
    }
    sol_count_++;
    return sol;
    
}

int
MotifStateSearch::finished() {
    if(sol_count_ >= max_solutions_ || no_more_solutions_) {
        if(verbose_) {
            std::cout << "maxed out solutions " << sol_count_ << " " << max_solutions_ <<  std::endl;
        }
        return 1;
    }
    else{
        return 0;
    }
}


MotifStateSearchSolutionOP
MotifStateSearch::_search() {
    float score;
    int i = 0;
    int steps = 0;
    int j = 0;
    int pos;
    MotifStateSearchNodeOP current, child, current_2;
    int clash;
    float dist = 1000;
    float best = 1000000000;
    MotifStateSearchSolutionOP best_sol;
    
    std::ofstream midpoint_file;
    auto midpoints = get_bool_option("save_midpoints");
    int midpoint_freq = get_int_option("save_midpoints_freq");
    
    if(midpoints) { midpoint_file.open(get_string_option("save_midpoints_file")); }
    
    while(! queue_.empty() ) {
        current = queue_.top();
        queue_.pop();
        
        
        steps += 1;
        
        if(steps > max_steps_) {
            if(verbose_) {
                std::cout << "MOTIF STATE SEARCH: reached max steps" << std::endl;
            }
            return best_sol;
        }
        
        if(midpoints && steps % midpoint_freq == 0) {
            auto sol = MotifStateSearchSolution(current, score);
            midpoint_file << sol.to_motif_tree()->topology_to_str() << std::endl;
            
        }

        score = scorer_->accept_score(current);

        if(score < best) {
            best = score;
            best_sol = std::make_shared<MotifStateSearchSolution>(current, score);

            if(verbose_) {
                std::cout << "MOTIF STATE SEARCH: best_score=" << best << " motifs_in_solution=";
                std::cout << current->level()-1 << " steps=" << steps << std::endl;
                
                //std::cout << best << " " << accept_score_ << " " << current->level() << " " << steps << " " << max_steps_ << std::endl;
            }
        }
        
        if(score < accept_score_ && current->ss_score() < min_ss_score_ &&
           current->level() > min_node_level_) {
            if(helix_end_ && current->ref_state()->name()[0] != 'H') {
                continue;
            }


            if(verbose_) {
                std::cout << "MOTIF STATE SEARCH: found a solution!" << std::endl;
            }
            auto s = std::make_shared<MotifStateSearchSolution>(current, score);
            return s;
        }
        
        if(current->level()+1 > max_node_level_) { continue; }

        possible_children_ = selector_->get_children_ms(current);
        pos = possible_children_.pos();
        test_node_->parent(current);
        Point center;
        i = -1;
        
        for(auto const & end : current->cur_state()->end_states()) {
            i++;
            if(i == 0) { continue; }
            j = -1;
            for(auto const & ms_and_type : possible_children_.motif_states_and_types()) {
                j++;
                
                if(j >= pos) { break; }
                //std::cout << current->level() << " " << current->ntype() << " " << ms_and_type.type << std::endl;
                test_node_->replace_ms(ms_and_type.motif_state,
                                       ms_and_type.type);
                
                
                if(current->size() + ms_and_type.motif_state->size() > max_size_) {
                    continue;
                }
                
                aligner_.get_aligned_motif_state(end,
                                                 test_node_->cur_state(),
                                                 test_node_->ref_state());
                test_node_->update(i);
                
                score = scorer_->score(test_node_);
                if(score > current->score()) { continue; }
                
                if(sterics_) {
                    clash = 0;
                    for(auto const & b1 : beads_) {
                        for(auto const & b2 : test_node_->cur_state()->beads()) {
                            dist = b1.distance(b2);
                            if(dist < 2.5) { clash = 1; break; }
                        }
                    }
                    if(clash) { continue; }
                    
                    if(using_lookup_) {
                        for(auto const & p : test_node_->cur_state()->beads()) {
                            //std::cout << p << " " << ms_and_type.motif_state->name() <<  std::endl;
                            clash = lookup_.clash(p);
                            if(clash) { break; }
                        }
                        if(clash) { continue; }
                    }
                    
                    current_2 = test_node_->parent();
                    while (current_2 != nullptr) {
                        for(auto const & b1 : test_node_->cur_state()->beads()) {
                            for(auto const & b2 : current_2->cur_state()->beads()) {
                                dist = b1.distance(b2);
                                if(dist < 2.5) { clash = 1; break; }
                            }
                            if(clash) { break; }
                        }
                        current_2 = current_2->parent();
                    }
                    if(clash) { continue; }

                }
                
             
                
                child = std::make_shared<MotifStateSearchNode>(*test_node_);
                child->update(i);
                child->score(score);
                queue_.push(child);
            }
        }
        
    }
    if(verbose_) {
        std::cout << "MOTIF STATE SEARCH: ran out of options" << std::endl;
    }
    
    if(midpoints) { midpoint_file.close(); }

    
    return best_sol;

}


MotifStateSearchNodeOP
MotifStateSearch::_start_node(
    BasepairStateOP const & start_bp) {
    
    auto ms = std::make_shared<MotifState>("start", Strings {"start", "start"},
                                           Strings {"", ""},
                                           BasepairStateOPs { start_bp, start_bp},
                                           Points(), 0, 0, 0);
    
    auto n = std::make_shared<MotifStateSearchNode>(ms, nullptr, -1, -1);
    n->setup_node_type_usage(selector_->size());
    
    return n;
}
